rmm/monitor/src/rmm/granule/entry.rs

```diff
@@ -15,16 +19,25 @@ use spinning_top::Spinlock;
 //  - Each entry can be either table (L1 table) or granule. This is determined by `Inner.table`.
 
 pub struct Granule {
+    /// granule state
     state: u64,
+    /// physical address which is aligned with GRANULE_SIZE
     addr: usize,
+    /// parent that this granule points to
+    /// the only case at this point is "Rd(parent) - Rec(child)"
+    /// Notice: do not put self-reference into this field, which may cause undefined behaviors.
+    parent: Option<Inner>,
 }
 
 impl Granule {
-    fn set_state(&mut self, addr: usize, state: u64) -> Result<(), Error> {
+    fn set_state<F>(&mut self, addr: usize, state: u64, destroy_callback: F) -> Result<(), Error>
+    where
+        F: Fn() -> Result<(), Error>,
+    {
         let prev = self.state;
 
         if prev == GranuleState::Undelegated {
-            if state != GranuleState::Delegated {
+            if state != GranuleState::Delegated && state != GranuleState::Undelegated {
```

Comment:
How about change this whole if-else statement to match?

```suggestion
let valid = match prev {
    GranuleState::Delegated => state != GranuleState::Delegated,
    _ => state == GranuleState::Delegated,
};

if !valid {
    return Err(Error::MmStateError);
}
```

Reply:
Reflected: